### PR TITLE
[SPARK-7112][Streaming][WIP] Add a InputInfoTracker to track all the input streams

### PR DESCRIPTION
**Pull request by jerryshao**

### CODE REVIEW

**Comment by tdas** [comment]
@jerryshao  This is a decent patch for the assumed design, but I mentioned in the parent JIRA https://issues.apache.org/jira/browse/SPARK-7111 that the design is not great. The "direct API" is a name we came up to differentiate new Kafka API from the old one receiver-based one, and logically every subclass of `InputDStream` that is not a `ReceiverInputDStream` is  a direct stream. So further separating out InputDStreams as direct stream and non-direct stream (beyond the receiver stream) is a bad idea. In the end, the goal is to enable all the streams, irrespective of its type, to report information about its input to the infra. And there should be a single common way / code path for doing it for all input streams, with some customization / override for receiver input streams (as unlike non-receiver-streams, all receiver streams has a common way of reporting block info).

I have been thinking about this since last night and I will post a very rough design doc on the JIRA shortly. Please follow up on the JIRA.

**Comment by tdas** [comment]
I have posted a different design on the JIRA. Please take a look.

**Comment by jerryshao** [comment]
Yes, I will take a look at the design doc, thanks a lot for your comments.

**Comment by tdas** [comment]
There are merge conflicts! Please merge master!

**Comment by jerryshao** [comment]
Yes, I will do this, please take take a look at the whole design, thanks a lot :)

File: streaming/src/main/scala/org/apache/spark/streaming/scheduler/InputInfoTracker.scala [hunk 1]
```diff
@@ -0,0 +1,74 @@
...
+case class InputInfo(batchTime: Time, inputStreamId: Int, numRecords: Long)
+
+/**
+ * This class manages all the input streams as well as their input data statistics. The information
+ * will output to StreamingListener to better monitoring.
+ */
+private[streaming] class InputInfoTracker(ssc: StreamingContext) extends Logging {
+
+  /** Track all the input streams registered in DStreamGraph */
+  val inputStreams = ssc.graph.getInputStreams()
```
Comment (tdas) [inline streaming/src/main/scala/org/apache/spark/streaming/scheduler/InputInfoTracker.scala]: Can this be private?

File: streaming/src/main/scala/org/apache/spark/streaming/scheduler/InputInfoTracker.scala [hunk 2]
```diff
@@ -0,0 +1,74 @@
...
+  /** Track all the id of input streams registered in DStreamGraph */
+  val inputStreamIds = inputStreams.map(_.id)
```
Comment (tdas) [inline streaming/src/main/scala/org/apache/spark/streaming/scheduler/InputInfoTracker.scala]: Can this be private?

File: streaming/src/main/scala/org/apache/spark/streaming/scheduler/InputInfoTracker.scala [hunk 3]
```diff
@@ -0,0 +1,74 @@
...
+
+  /** Get the all the input stream's information of specified batch time */
+  def getInfo(batchTime: Time): Map[Int, InputInfo] = synchronized {
+    val inputInfos = batchTimeToInputInfos.get(batchTime)
+    // Convert mutable HashMap to immutable Map for the caller
+    inputInfos.map(_.toMap).getOrElse(Map[Int, InputInfo]())
+  }
+
+  /** Get the input information of specified batch time and input stream id */
+  def getInfoOfBatchAndStream(batchTime: Time, inputStreamId: Int
```
Comment (tdas) [inline streaming/src/main/scala/org/apache/spark/streaming/scheduler/InputInfoTracker.scala]: This is not used anywhere other than tests, is this necessary?

File: streaming/src/main/scala/org/apache/spark/streaming/scheduler/JobScheduler.scala [hunk 4]
```diff
@@ -50,6 +50,8 @@ class JobScheduler(val ssc: StreamingContext) extends Logging {
   // These two are created only when scheduler starts.
   // eventLoop not being null means the scheduler has been started and not stopped
   var receiverTracker: ReceiverTracker = null
+  // A tracker to track all the input stream information as well as processed record number
+  var inputInfoTracker: InputInfoTracker = null
```
Comment (tdas) [inline streaming/src/main/scala/org/apache/spark/streaming/scheduler/JobScheduler.scala]: an empty line after this would be good.

File: streaming/src/main/scala/org/apache/spark/streaming/scheduler/BatchInfo.scala [hunk 5]
```diff
@@ -32,7 +33,7 @@ import org.apache.spark.streaming.Time
 @DeveloperApi
 case class BatchInfo(
     batchTime: Time,
-    receivedBlockInfo: Map[Int, Array[ReceivedBlockInfo]],
+    streamIdToNumRecords: Map[Int, Long],
     submissionTime: Long,
     processingStartTime: Option[Long],
     processingEndTime: Option[Long]
```
Comment (tdas) [inline streaming/src/main/scala/org/apache/spark/streaming/scheduler/BatchInfo.scala]: Can you make a method called `numRecords` which returns the sum? This is the same approach taken by @zsxwing in  #5533, so will be easier to merge conflicts later.

File: streaming/src/main/scala/org/apache/spark/streaming/ui/StreamingJobProgressListener.scala [hunk 6]
```diff
@@ -70,9 +70,8 @@ private[streaming] class StreamingJobProgressListener(ssc: StreamingContext)
     runningBatchInfos(batchStarted.batchInfo.batchTime) = batchStarted.batchInfo
     waitingBatchInfos.remove(batchStarted.batchInfo.batchTime)
 
-    batchStarted.batchInfo.receivedBlockInfo.foreach { case (_, infos) =>
-      totalReceivedRecords += infos.map(_.numRecords).sum
-    }
+    // TODO. this should be fixed when input stream is not receiver based stream.
+    totalReceivedRecords += batchStarted.batchInfo.streamIdToNumRecords.values.sum
```
Comment (tdas) [inline streaming/src/main/scala/org/apache/spark/streaming/ui/StreamingJobProgressListener.scala]: This can be replaced by `batchStarted.batchInfo.numRecords` if you implement `numRecords` as I said above.

File: streaming/src/main/scala/org/apache/spark/streaming/ui/StreamingJobProgressListener.scala [hunk 7]
```diff
@@ -135,28 +132,25 @@ private[streaming] class StreamingJobProgressListener(ssc: StreamingContext)
...
-        batchInfo.get(receiverId).getOrElse(Array.empty)
-      }
-      val recordsOfParticularReceiver = blockInfoOfParticularReceiver.map { blockInfo =>
-      // calculate records per second for each batch
-        blockInfo.map(_.numRecords).sum.toDouble * 1000 / batchDuration
-      }
-      val distributionOption = Distribution(recordsOfParticularReceiver)
-      (receiverId, distributionOption)
+
+    // TODO. this should be fixed when receiver-less input stream is mixed into BatchInfo
```
Comment (tdas) [inline streaming/src/main/scala/org/apache/spark/streaming/ui/StreamingJobProgressListener.scala]: What does this to do mean?

File: streaming/src/main/scala/org/apache/spark/streaming/scheduler/InputInfoTracker.scala (hunk 3 above)
Comment (jerryshao) [inline streaming/src/main/scala/org/apache/spark/streaming/scheduler/InputInfoTracker.scala]: yes, only used for test, I can remove it if necessary.

File: streaming/src/main/scala/org/apache/spark/streaming/ui/StreamingJobProgressListener.scala (hunk 7 above)
Comment (jerryshao) [inline streaming/src/main/scala/org/apache/spark/streaming/ui/StreamingJobProgressListener.scala]: This is what makes me concern a lot. Now for the `BatchInfo's streamIdToNumRecords`, all the input stream's statistic data will be in it, not receiver-based input stream, so do we need to differentiate the statistics?

File: streaming/src/main/scala/org/apache/spark/streaming/ui/StreamingJobProgressListener.scala (hunk 6 above)
Comment (jerryshao) [inline streaming/src/main/scala/org/apache/spark/streaming/ui/StreamingJobProgressListener.scala]: Yes, will do. Also have one concern, if the `batchStarted` is not a receiver-based batchInfo, so do we need to count this records into `totalReceivedRecords` when batch is just started.

File: streaming/src/main/scala/org/apache/spark/streaming/ui/StreamingPage.scala [hunk 8]
```diff
@@ -95,7 +95,7 @@ private[ui] class StreamingPage(parent: StreamingTab)
         "Maximum rate\n[events/sec]",
         "Last Error"
       )
-      val dataRows = (0 until listener.numReceivers).map { receiverId =>
```
Comment (jerryshao) [inline streaming/src/main/scala/org/apache/spark/streaming/ui/StreamingPage.scala]: Now all the input streams will have a unique id (not only receiver based input streams), so assuming  this will get error.

File: streaming/src/main/scala/org/apache/spark/streaming/scheduler/InputInfoTracker.scala (hunk 2 above)
Comment (jerryshao) [inline streaming/src/main/scala/org/apache/spark/streaming/scheduler/InputInfoTracker.scala]: Originally I'd like to expose these, but these can also be gotten from DStreamGraph, so I will remove these.

**Comment by jerryshao** [comment]
Hi @zsxwing, you also changed a lot on this part, would you please take a look at this, thanks a lot.

**Comment by jerryshao** [comment]
Jenkins, retest this please.

File: streaming/src/main/scala/org/apache/spark/streaming/ui/StreamingJobProgressListener.scala [hunk 9]
```diff
@@ -101,6 +101,7 @@ private[streaming] class StreamingJobProgressListener(ssc: StreamingContext)
     runningBatchUIData(batchStarted.batchInfo.batchTime) = BatchUIData(batchStarted.batchInfo)
     waitingBatchUIData.remove(batchStarted.batchInfo.batchTime)
 
+    // TODO. this should be fixed when input stream is not receiver based stream.
```
Comment (jerryshao) [inline streaming/src/main/scala/org/apache/spark/streaming/ui/StreamingJobProgressListener.scala]: All these TODOs means that: now `BatchInfo#streamIdtoNumRecords` will track both the input stream's received data (for receiver-based stream) and to-be-processed data number (for receiver-less stream). So the semantics in `StreamingJobProgressListener` is slightly different.

Take `totalReceivedRecords` as an example: should we add the receiver-less input stream's number of to-be-processed record into here, since it is not actually received yet. So we have to figure out this semantics difference.

File: streaming/src/main/scala/org/apache/spark/streaming/ui/StreamingJobProgressListener.scala [hunk 10]
```diff
@@ -188,25 +189,26 @@ private[streaming] class StreamingJobProgressListener(ssc: StreamingContext)
...
-    (0 until numReceivers).map { receiverId =>
-      val recordsOfParticularReceiver = latestBatches.map { batch =>
-        // calculate records per second for each batch
-        batch.receiverNumRecords.get(receiverId).sum.toDouble * 1000 / batchDuration
-      }
-      val distributionOption = Distribution(recordsOfParticularReceiver)
-      (receiverId, distributionOption)
+    val latestBatchInfos = retainedBatches.reverse.take(batchUIDataLimit)
+
+    // TODO. this should be fixed when receiver-less input stream is mixed into BatchInfo
```
Comment (jerryshao) [inline streaming/src/main/scala/org/apache/spark/streaming/ui/StreamingJobProgressListener.scala]: Maybe this TODO is not necessary.

File: streaming/src/main/scala/org/apache/spark/streaming/scheduler/InputInfoTracker.scala [hunk 11]
```diff
@@ -0,0 +1,69 @@
...
+
+import org.apache.spark.Logging
+import org.apache.spark.streaming.{Time, StreamingContext}
+
+/** To track the information of input stream at specified batch time. */
+case class InputInfo(batchTime: Time, inputStreamId: Int, numRecords: Long)
+
+/**
+ * This class manages all the input streams as well as their input data statistics. The information
+ * will output to StreamingListener to better monitoring.
```
Comment (zsxwing) [inline streaming/src/main/scala/org/apache/spark/streaming/scheduler/InputInfoTracker.scala]: "to better monitoring" -> "for better monitoring"

File: streaming/src/main/scala/org/apache/spark/streaming/ui/StreamingJobProgressListener.scala (hunk 6 above)
Comment (tdas) [inline streaming/src/main/scala/org/apache/spark/streaming/ui/StreamingJobProgressListener.scala]: Yes we need to. Receiver or not, all the data has been received by the
system.

File: streaming/src/main/scala/org/apache/spark/streaming/scheduler/InputInfoTracker.scala (hunk 11 above)
Comment (jerryshao) [inline streaming/src/main/scala/org/apache/spark/streaming/scheduler/InputInfoTracker.scala]: Done :)

**Comment by jerryshao** [comment]
Any comment for the current implementation ? @tdas

File: streaming/src/main/scala/org/apache/spark/streaming/scheduler/InputInfoTracker.scala (hunk 3 above)
Comment (tdas) [inline streaming/src/main/scala/org/apache/spark/streaming/scheduler/InputInfoTracker.scala]: Please remove it if it is not needed by anything non-test.

Comment (jerryshao) [inline streaming/src/main/scala/org/apache/spark/streaming/scheduler/InputInfoTracker.scala]: Done.

File: streaming/src/main/scala/org/apache/spark/streaming/ui/StreamingJobProgressListener.scala (hunk 9 above)
Comment (tdas) [inline streaming/src/main/scala/org/apache/spark/streaming/ui/StreamingJobProgressListener.scala]: That's a good point. I think it is okay to add it for now. At least under normal operation, when things are keeping up, it is useful to see what is the rate. If it does not keep up, then there are other indicators.

File: external/kafka/src/main/scala/org/apache/spark/streaming/kafka/DirectKafkaInputDStream.scala [hunk 12]
```diff
@@ -17,7 +17,6 @@
 
 package org.apache.spark.streaming.kafka
 
-
```
Comment (tdas) [inline external/kafka/src/main/scala/org/apache/spark/streaming/kafka/DirectKafkaInputDStream.scala]: Please revert this file.

File: streaming/src/main/scala/org/apache/spark/streaming/scheduler/InputInfoTracker.scala [hunk 13]
```diff
@@ -0,0 +1,63 @@
...
+
+import org.apache.spark.Logging
+import org.apache.spark.streaming.{Time, StreamingContext}
+
+/** To track the information of input stream at specified batch time. */
+case class InputInfo(batchTime: Time, inputStreamId: Int, numRecords: Long)
+
+/**
+ * This class manages all the input streams as well as their input data statistics. The information
+ * will output to StreamingListener for better monitoring.
```
Comment (tdas) [inline streaming/src/main/scala/org/apache/spark/streaming/scheduler/InputInfoTracker.scala]: `will be exposed through StreamingListener for monitoring.`

File: streaming/src/main/scala/org/apache/spark/streaming/ui/StreamingJobProgressListener.scala (hunk 9 above)
Comment (jerryshao) [inline streaming/src/main/scala/org/apache/spark/streaming/ui/StreamingJobProgressListener.scala]: OK, I will keep as it is for now and remove all these TODOs. When enable this in `KafkaDirectInputDStream`, we could rethink about what exact semantics should be in this part.

File: streaming/src/main/scala/org/apache/spark/streaming/scheduler/InputInfoTracker.scala [hunk 14]
```diff
@@ -0,0 +1,63 @@
...
+
+package org.apache.spark.streaming.scheduler
+
+import scala.collection.mutable
+
+import org.apache.spark.Logging
+import org.apache.spark.streaming.{Time, StreamingContext}
+
+/** To track the information of input stream at specified batch time. */
+case class InputInfo(batchTime: Time, inputStreamId: Int, numRecords: Long)
```
Comment (tdas) [inline streaming/src/main/scala/org/apache/spark/streaming/scheduler/InputInfoTracker.scala]: This has to be made private[scheduler] or private[streaming]

**Comment by tdas** [comment]
The approach is good! I am still looking at the details. There are some details that are going to to be changed singificantly in #5533 by @zsxwing . So I will look for general correctness.

File: streaming/src/test/scala/org/apache/spark/streaming/scheduler/InputInfoTrackerSuite.scala [hunk 15]
```diff
@@ -0,0 +1,104 @@
...
+      ssc.stop()
+      ssc = null
+    }
+  }
+
+  test("test track the number of input stream") {
+    class TestInputDStream extends InputDStream[String](ssc) {
+      def start() { }
+      def stop() { }
+      def compute(validTime: Time): Option[RDD[String]] = ???
```
Comment (tdas) [inline streaming/src/test/scala/org/apache/spark/streaming/scheduler/InputInfoTrackerSuite.scala]: Do we use this convention in Spark? I dont think so. Please replace with { }

File: streaming/src/test/scala/org/apache/spark/streaming/scheduler/InputInfoTrackerSuite.scala [hunk 16]
```diff
@@ -0,0 +1,104 @@
...
+    }
+
+    class TestReceiverInputDStream extends ReceiverInputDStream[String](ssc) {
+      def getReceiver: Receiver[String] = ???
```
Comment (tdas) [inline streaming/src/test/scala/org/apache/spark/streaming/scheduler/InputInfoTrackerSuite.scala]: Please replace with null.

File: streaming/src/main/scala/org/apache/spark/streaming/scheduler/InputInfoTracker.scala (hunk 14 above)
Comment (tdas) [inline streaming/src/main/scala/org/apache/spark/streaming/scheduler/InputInfoTracker.scala]: Is the batchTime needed in this? Isn't just inputStreamId and numberRecords fine?

File: streaming/src/test/scala/org/apache/spark/streaming/scheduler/InputInfoTrackerSuite.scala [hunk 17]
```diff
@@ -0,0 +1,104 @@
...
+  }
+
+  after {
+    if (ssc != null) {
+      ssc.stop()
+      ssc = null
+    }
+  }
+
+  test("test track the number of input stream") {
```
Comment (tdas) [inline streaming/src/test/scala/org/apache/spark/streaming/scheduler/InputInfoTrackerSuite.scala]: This is test is unrelated to the InputInfoTracker, so probably should not be in here. Rather put it in `InputStreamsSuite`.

File: streaming/src/main/scala/org/apache/spark/streaming/scheduler/InputInfoTracker.scala (hunk 14 above)
Comment (jerryshao) [inline streaming/src/main/scala/org/apache/spark/streaming/scheduler/InputInfoTracker.scala]: Yeah, actually `batchTime` is not use in anywhere, I will remove this.

File: streaming/src/test/scala/org/apache/spark/streaming/scheduler/InputInfoTrackerSuite.scala [hunk 18]
```diff
@@ -0,0 +1,104 @@
...
+    inputInfoTracker.reportInfo(inputInfo2.batchTime, inputInfo2)
+
+    inputInfoTracker.cleanup(Time(0))
+    assert(inputInfoTracker.getInfo(Time(0))(streamId1) === inputInfo1)
+    assert(inputInfoTracker.getInfo(Time(1))(streamId1) === inputInfo2)
+
+    inputInfoTracker.cleanup(Time(1))
+    assert(inputInfoTracker.getInfo(Time(0)).get(streamId1) === None)
+    assert(inputInfoTracker.getInfo(Time(1))(streamId1) === inputInfo2)
+  }
```
Comment (tdas) [inline streaming/src/test/scala/org/apache/spark/streaming/scheduler/InputInfoTrackerSuite.scala]: I think we need an end-to-end test as well, to test the code path interacting between JobGenerator, StreamingListener and InputInfoTracker. I thought about what is a simple way to do and here is an idea. Checkpoint StreamingListenerSuite. The first test "batch info" can be extended to test the numRecords. For that to work, the input stream used in these tests is [TestInputDStream](https://github.com/apache/spark/blob/master/streaming/src/test/scala/org/apache/spark/streaming/TestSuiteBase.scala#L43). You could extend add a line in the TestSuiteBased.compute to post InputInfo. And then test the info in StreamingListenerSuite. Does that make sense?

**Comment by tdas** [comment]
I made some more comments. Would be great if you can get it done today and I can merge it in tomorrow. That will unblock #5533.

**Comment by jerryshao** [comment]
OK, I will try to get it done, thanks a lot for your comments.

File: streaming/src/main/scala/org/apache/spark/streaming/scheduler/InputInfoTracker.scala [hunk 19]
```diff
@@ -0,0 +1,63 @@
...
+    val inputInfos = batchTimeToInputInfos.get(batchTime)
+    // Convert mutable HashMap to immutable Map for the caller
+    inputInfos.map(_.toMap).getOrElse(Map[Int, InputInfo]())
+  }
+
+  /** Cleanup the tracked input information older than threshold batch time */
+  def cleanup(batchThreshTime: Time): Unit = synchronized {
+    val timesToCleanup = batchTimeToInputInfos.keys.filter(_ < batchThreshTime)
+    logInfo(s"remove old batch metadata: ${timesToCleanup.mkString(" ")}")
+
```
Comment (tdas) [inline streaming/src/main/scala/org/apache/spark/streaming/scheduler/InputInfoTracker.scala]: Can remove this empty line.

File: streaming/src/test/scala/org/apache/spark/streaming/scheduler/InputInfoTrackerSuite.scala (hunk 18 above)
Comment (jerryshao) [inline streaming/src/test/scala/org/apache/spark/streaming/scheduler/InputInfoTrackerSuite.scala]: Hi @tdas , what's the meaning of `Checkpoint StreamingListenerSuite`?

Comment (tdas) [inline streaming/src/test/scala/org/apache/spark/streaming/scheduler/InputInfoTrackerSuite.scala]: Sorry, typo. I meant "Checkout StreamingListenerSuite."

:)

File: streaming/src/main/scala/org/apache/spark/streaming/ui/StreamingJobProgressListener.scala [hunk 20]
```diff
@@ -188,25 +188,26 @@ private[streaming] class StreamingJobProgressListener(ssc: StreamingContext)
...
-        // calculate records per second for each batch
-        batch.receiverNumRecords.get(receiverId).sum.toDouble * 1000 / batchDuration
-      }
-      val distributionOption = Distribution(recordsOfParticularReceiver)
-      (receiverId, distributionOption)
+    val latestBatchInfos = retainedBatches.reverse.take(batchUIDataLimit)
+    val latestReceiverNumRecords = latestBatchInfos.map(_.receiverNumRecords)
+    val streamIds = ssc.graph.getInputStreams().map(_.id)
+    streamIds.map { id =>
+     val recordsOfParticularReceiver =
```
Comment (tdas) [inline streaming/src/main/scala/org/apache/spark/streaming/ui/StreamingJobProgressListener.scala]: incorrect indentation.

Comment (tdas) [inline streaming/src/main/scala/org/apache/spark/streaming/ui/StreamingJobProgressListener.scala]: @zsxwing Can you take care of this indentation in your PR?

**Comment by tdas** [comment]
LGTM. Merging this. Thanks!

File: streaming/src/main/scala/org/apache/spark/streaming/ui/StreamingJobProgressListener.scala (hunk 20 above)
Comment (jerryshao) [inline streaming/src/main/scala/org/apache/spark/streaming/ui/StreamingJobProgressListener.scala]: Sorry @tdas for late response, I just went to the hospital this morning. @zsxwing , would you please help me to fix this, thanks a lot.
